src/PowerShellEditorServices/Session/InvocationEventQueue.cs

```diff
@@ -0,0 +1,251 @@
+            StringBuilder errorMessages,
+            ExecutionOptions executionOptions)
+        {
+            var request = new PipelineExecutionRequest<TResult>(
+                _powerShellContext,
+                psCommand,
+                errorMessages,
+                executionOptions);
+
+            await SetInvocationRequestAsync(
+                new InvocationRequest(
+                    pwsh => request.Execute().GetAwaiter().GetResult()));
+
+            try
+            {
+                return await request.Results;
+            }
+            finally
+            {
+                await SetInvocationRequestAsync(null);
+            }
+        }
+
+        /// <summary>
+        /// Marshals a <see cref="Action{PowerShell}" /> to run on the pipeline thread. A new
+        /// <see cref="PromptNestFrame" /> will be created for the invocation.
+        /// </summary>
+        /// <param name="invocationAction">
+        /// The <see cref="Action{PowerShell}" /> to invoke on the pipeline thread. The nested
+        /// <see cref="PowerShell" /> instance for the created <see cref="PromptNestFrame" />
+        /// will be passed as an argument.
+        /// </param>
+        /// <returns>
+        /// An awaitable <see cref="Task" /> that the caller can use to know when execution completes.
+        /// </returns>
+        internal async Task InvokeOnPipelineThread(Action<PowerShell> invocationAction)
+        {
+            var request = new InvocationRequest(pwsh =>
+            {
+                using (_promptNest.GetRunspaceHandle(CancellationToken.None, isReadLine: false))
+                {
+                    pwsh.Runspace = _runspace;
+                    invocationAction(pwsh);
+                }
+            });
+
+            await SetInvocationRequestAsync(request);
+            try
+            {
+                await request.Task;
+            }
+            finally
+            {
+                await SetInvocationRequestAsync(null);
+            }
+        }
+
+        private async Task WaitForExistingRequestAsync()
+        {
+            InvocationRequest existingRequest;
+            await _lock.WaitAsync();
+            try
+            {
+                existingRequest = _invocationRequest;
+                if (existingRequest == null || existingRequest.Task.IsCompleted)
+                {
+                    return;
+                }
+            }
+            finally
+            {
+                _lock.Release();
+            }
+
+            await existingRequest.Task;
+        }
+
+        private async Task SetInvocationRequestAsync(InvocationRequest request)
+        {
+            await WaitForExistingRequestAsync();
+            await _lock.WaitAsync();
+            try
+            {
+                _invocationRequest = request;
+            }
+            finally
+            {
+                _lock.Release();
+            }
+
+            _powerShellContext.ForcePSEventHandling();
```

Comment:
So while PSReadLine is running it has full control of the pipeline.  While it's waiting for a key it will timeout once every 300 milliseconds and check for PowerShell events to process.  When it does that, we use the subscriber to take over the pipeline thread.  `ForcePSEventHandling` is a "private contract" method in PSRL that skips the 300 millisecond wait and goes straight to event processing.